llvm/lib/Target/AArch64/AArch64RegisterInfo.cpp

```diff
@@ -424,6 +424,58 @@ AArch64RegisterInfo::explainReservedReg(const MachineFunction &MF,
   return {};
 }
 
+static SmallVector<MCPhysReg> ReservedHi = {
```

Comment:
What kind of failures? AMDGPU also has synthetic 16-bit high sub registers and they are not explicitly reserved. Are you adding these to an allocatable class?

Reply:
But what are the actual failures, messages, location?  If the high half of register isn't allocatable / addressable in the first place, it shouldn't just appear to cause issues

Reply:
Without marking the registers as reserved, then for the example below:
```
---
name:            sv2i64
tracksRegLiveness: true
body:             |
  bb.0.entry:
    liveins: $q0, $q1

    %0:fpr128 = COPY $q0
    %1:fpr128 = COPY $q1
    %35:gpr64 = COPY %0.dsub
    %36:gpr64 = COPY %1.dsub
    %9:gpr64 = SDIVXr %35, %36
    %37:gpr64 = UMOVvi64 %0, 1
    %38:gpr64 = UMOVvi64 %1, 1
    %10:gpr64 = SDIVXr %37, %38
    %19:fpr128 = INSvi64gpr undef %19, 0, %9
    %19:fpr128 = INSvi64gpr %19, 1, %10
    %39:gpr64 = COPY %19.dsub
    %24:gpr64 = MADDXrrr %39, %36, $xzr
    %41:gpr64 = UMOVvi64 %19, 1
    %25:gpr64 = MADDXrrr %41, %38, $xzr
    %34:fpr128 = INSvi64gpr undef %34, 0, %24
    %34:fpr128 = INSvi64gpr %34, 1, %25
    %2:fpr128 = SUBv2i64 %0, %34
    $q0 = COPY %2
    RET_ReallyLR implicit $q0
...
```

When I run this with:
```
llc -global-isel -verify-machineinstrs -run-pass=machine-scheduler
```

It fails with:
```
Use of $xzr does not have a corresponding definition on every path:
216r %10:gpr64 = MADDXrrr %9:gpr64, %3:gpr64, $xzr
LLVM ERROR: Use not jointly dominated by defs.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: ./bin/llc -global-isel -verify-machineinstrs -run-pass=machine-scheduler /tmp/t.mir -o -
1.      Running pass 'Function Pass Manager' on module '/tmp/t.mir'.
2.      Running pass 'Machine Instruction Scheduler' on function '@sv2i64'
 ...
 #8 0x0000ffff80062b7c llvm::LiveRangeCalc::findReachingDefs(llvm::LiveRange&, llvm::MachineBasicBlock&, llvm::SlotIndex, unsigned int, llvm::ArrayRef<llvm::SlotIndex>)
 #9 0x0000ffff80063e94 llvm::LiveRangeCalc::extend(llvm::LiveRange&, llvm::SlotIndex, unsigned int, llvm::ArrayRef<llvm::SlotIndex>)
#10 0x0000ffff80064a18 llvm::LiveIntervalCalc::extendToUses(llvm::LiveRange&, llvm::Register, llvm::LaneBitmask, llvm::LiveInterval*)
#11 0x0000ffff8003e82c llvm::LiveIntervals::computeRegUnitRange(llvm::LiveRange&, unsigned int)
#12 0x0000ffff80044cdc llvm::LiveIntervals::HMEditor::updateAllRanges(llvm::MachineInstr*)
#13 0x0000ffff8004848c llvm::LiveIntervals::handleMove(llvm::MachineInstr&, bool)
#14 0x0000ffff801f44ec llvm::ScheduleDAGMI::moveInstruction(llvm::MachineInstr*, llvm::MachineInstrBundleIterator<llvm::MachineInstr, false>)
#15 0x0000ffff801fdb58 llvm::ScheduleDAGMILive::scheduleMI(llvm::SUnit*, bool)
#16 0x0000ffff8020b214 llvm::ScheduleDAGMILive::schedule()
#17 0x0000ffff801f0934 (anonymous namespace)::MachineSchedulerBase::scheduleRegions(llvm::ScheduleDAGInstrs&, bool) (.isra.0) MachineScheduler.cpp:0:0
```

Reply:
This smells like an unrelated bug, this is not the kind of error I expected

Reply:
I don't think there is a bug; the code for moving an instruction goes through the list of operands to update the register's liverange. For each physreg it then goes through the regunits to calculate/update the liverange for that regunit, but only if the regunit is not reserved.

The code that determines if the register is reserved says:
```
// A register unit is considered reserved if all its roots and all their
// super registers are reserved.
```
Without this change to AArch64RegisterInfo.cpp, WZR and XZR are marked as reserved, but WZR_HI isn't (because WZR_HI is a sibling of WZR, and `markSuperRegs` marks only XZR as reserved), and so `IsReserved` is `false` for the WZR_HI regunit.

Why this doesn't fail for AMDGPU I don't know, perhaps these registers are always virtual and they never go down this path.